### PR TITLE
Enable bootstrap customization of flux components

### PR DESCRIPTION
**Pull request by rtjfarrimond**
### Overview

This PR enables users of `terraform-provider-flux` to customize the flux installation in the same manner as recorded in the existing flux documentation for manual configuration ([here](https://fluxcd.io/docs/installation/#customize-flux-manifests)), for tasks such as configuring sops using a cloud KMS.

This change improves upon the currently [documented approach](https://github.com/fluxcd/terraform-provider-flux/blob/main/docs/guides/customize-flux.md) to customize the installation, by removing the need for the user to have to create their own `kustomization-override.yaml`. This ensures that responsibility of generating `kustomize.yaml` remains with the terraform provider, meaning users do not risk missing out on any updates that may be made to the default `kustomize.yaml` generated by `flux_sync` in future releases by overriding it with their own yaml file.

This behaviour is achieved by allowing users to pass in a set of names of patches (`patch_names`) that they will use to customize the installation to `flux_sync`, which will return as an output value (`patch_file_paths`) a map whose key set contains only the values passed in `patch_names` with corresponding values detailing the path in github where flux expect these patch files to be uploaded. The user of the provider then has the responsibility of creating the patch files at the paths expected by flux. Additionally, `flux_sync` will include in `kustomization.yaml` a `patchesStrategicMerge` block with a list of all patches that were passed into `patch_names`, following the naming convention of `patch-<patch-name>.yaml`. Enabling the user to provide many patch files in this manner allows them to follow the [Kustomize recommendation](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#customizing) of declaring 'small patches that do one thing'.

Fix: #120 

Signed-off-by: Robert Farrimond <robert.farrimond@ovoenergy.com>

### CODE REVIEW

File: tools.go [hunk 1]
```diff
@@ -1,3 +1,4 @@
+//go:build tools
```
Comment (phillebaba) [inline tools.go:1]: We have not upgraded to go 1.17 yet.

File: docs/guides/customize-flux.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  type = string
+}
+
+data "google_service_account" "flux_sops" {
+  account_id = "flux-sops"
+}
+
+locals {
+  // 'Small patches that do one thing are recommended'
+ //    - https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#customizing
```
Comment (phillebaba) [inline docs/guides/customize-flux.md:83]: Use `#` for comments

**Review by phillebaba — COMMENTED** [review]
Left some smaller syntax comments, but I think I have a bigger question we need to answer.

So this PR seems to solve the Kustomization template generation, but not the issue described in #120. There are two methods of customizing flux bootstrap one is you do it before and the other is after. This will still only to the customizing after the initial bootstrap as it does not modify the manifests generated that are applied. Or have I missed an important detail?

**Comment by stefanprodan** [comment]
I don't see how this could solve the customisations at install time. To install the components along with their patches, TF should implement a staged server-side apply like we do in the CLI https://github.com/fluxcd/flux2/blob/main/internal/bootstrap/bootstrap_plain_git.go#L172-L190

**Comment by rtjfarrimond** [comment]
@stefanprodan is your concern that you don't think that this would work, or that the implementation should be different? If the former, then I can confirm that I have tested this using a GKE cluster and the patches provided are applied to the flux installation.

**Comment by rtjfarrimond** [comment]
@phillebaba 
> it does not modify the manifests generated that are applied
Do you mean the gotk manifests? No they are not modified, but the patches defined in `kustomize.yaml` are applied to extend / override them. The set of patches passed to the provider can be modified after the initial install to add or remove patches as needed.

**Comment by stefanprodan** [comment]
@rtjfarrimond this PR looks Ok to me, sorry I was replying to @phillebaba.

File: docs/data-sources/sync.md [hunk 3]
```diff
@@ -44,6 +44,7 @@ data "flux_sync" "main" {
 - **interval** (Number) Sync interval in minutes. Defaults to `1`.
 - **name** (String) The kubernetes resources name Defaults to `flux-system`.
 - **namespace** (String) The namespace scope for this operation. Defaults to `flux-system`.
+- **patch_names** (Set of String) The names of patches to apply to the initial Kustomization. Used to generate the `patch_file_paths` output value.
```
Comment (phillebaba) [inline docs/data-sources/sync.md:47]: The use of the word "initial" will probably create confusion.

File: docs/guides/customize-flux.md [hunk 4]
```diff
@@ -7,18 +7,17 @@ description: |-
...
-exposed by the `flux bootstrap` CLI command. There may be situations where the exposed parameters are not enough and additional configuration
+The Flux datasources expose a set of parameters that can configure the controller deployment. These parameters include all of those
+exposed by the `flux bootstrap` CLI command. There may be situations where these parameters are not enough and additional configuration
 has to be done. This could be changing the resource requests or limits for a controller, adding annotations or labels, or modifying container settings.
 
 When deploying Flux with the CLI the recommended solution is to [modify the Kustomization file](https://fluxcd.io/docs/installation/#customize-flux-manifests).
-This is also the recommended way when deploying with Terraform, but it requires some extra considerations as changes made manually to
-the repository after deployment would be overridden the next time the Terraform is applied.
+When using terraform to bootstrap flux, the `flux_sync` data source provides a `patch_names` argument and corresponding `patch_file_paths` output value that allow users to provide patches to the initial `kustomize.yaml` in the same manner.
```
Comment (phillebaba) [inline docs/guides/customize-flux.md:15]: Ditto initial

**Comment by phillebaba** [comment]
@rtjfarrimond two small comments and a rebase and then we should be able to merge.

**Comment by rtjfarrimond** [comment]
Thanks @phillebaba, removed 'initial' and rebased.

**Review by phillebaba — APPROVED** [review]
LGTM great job! :sunflower:

**Pull request merged**